File: GooglePay/src/main/java/com/braintreepayments/api/googlepay/GooglePayInternalClient.java
package com.braintreepayments.api.googlepay;

import android.content.Context;
<<<<<<< HEAD:GooglePay/src/main/java/com/braintreepayments/api/googlepay/GooglePayInternalClient.java
=======

import androidx.annotation.NonNull;
>>>>>>> main:GooglePay/src/main/java/com/braintreepayments/api/GooglePayInternalClient.java

import com.braintreepayments.api.core.Configuration;
import com.google.android.gms.common.api.ApiException;
import com.google.android.gms.wallet.IsReadyToPayRequest;
import com.google.android.gms.wallet.PaymentsClient;
import com.google.android.gms.wallet.Wallet;
import com.google.android.gms.wallet.WalletConstants;

class GooglePayInternalClient {

    void isReadyToPay(Context context, Configuration configuration, IsReadyToPayRequest isReadyToPayRequest, final GooglePayIsReadyToPayCallback callback) {
        PaymentsClient paymentsClient = Wallet.getPaymentsClient(context,
                new Wallet.WalletOptions.Builder()
                        .setEnvironment(getGooglePayEnvironment(configuration))
                        .build());
        paymentsClient.isReadyToPay(isReadyToPayRequest).addOnCompleteListener(task -> {
            try {
                Boolean isReady = task.getResult(ApiException.class);
                if (isReady) {
                    callback.onGooglePayReadinessResult(GooglePayReadinessResult.ReadyToPay.INSTANCE);
                } else {
                    callback.onGooglePayReadinessResult(new GooglePayReadinessResult.NotReadyToPay(null));
                }
            } catch (ApiException e) {
                callback.onGooglePayReadinessResult(new GooglePayReadinessResult.NotReadyToPay(e));
            }
        });
    }

    int getGooglePayEnvironment(Configuration configuration) {
        if ("production".equals(configuration.getGooglePayEnvironment())) {
            return WalletConstants.ENVIRONMENT_PRODUCTION;
        } else {
            return WalletConstants.ENVIRONMENT_TEST;
        }
    }
}

File: GooglePay/src/main/java/com/braintreepayments/api/googlepay/GooglePayInternalClient.kt
package com.braintreepayments.api.googlepay

import android.content.Context
import com.braintreepayments.api.core.Configuration
import com.braintreepayments.api.googlepay.GooglePayReadinessResult.NotReadyToPay
import com.google.android.gms.common.api.ApiException
import com.google.android.gms.tasks.Task
import com.google.android.gms.wallet.IsReadyToPayRequest
import com.google.android.gms.wallet.Wallet
import com.google.android.gms.wallet.Wallet.WalletOptions
import com.google.android.gms.wallet.WalletConstants
import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine

internal class GooglePayInternalClient {
    suspend fun isReadyToPay(
        context: Context,
        configuration: Configuration,
        isReadyToPayRequest: IsReadyToPayRequest
    ): GooglePayReadinessResult {
        return suspendCoroutine { continuation ->
            val paymentsClient = Wallet.getPaymentsClient(
                context,
                WalletOptions.Builder()
                    .setEnvironment(getGooglePayEnvironment(configuration))
                    .build()
            )
            paymentsClient.isReadyToPay(isReadyToPayRequest)
                .addOnCompleteListener { task: Task<Boolean> ->
                    try {
                        val isReady = task.getResult(ApiException::class.java)
                        if (isReady) {
                            continuation.resume(GooglePayReadinessResult.ReadyToPay)
                        } else {
                            continuation.resume(NotReadyToPay(null))
                        }
                    } catch (e: ApiException) {
                        continuation.resume(NotReadyToPay(e))
                    }
                }
        }
    }

    private fun getGooglePayEnvironment(configuration: Configuration): Int {
        return if ("production" == configuration.googlePayEnvironment) {
            WalletConstants.ENVIRONMENT_PRODUCTION
        } else {
            WalletConstants.ENVIRONMENT_TEST
        }
    }
}
